scripts/validate_string_concatenation.py

```diff
@@ -1,51 +1,313 @@
+                yield (
+                    source_path,
+                    current_token.start[0],
+                    "Bare pytests raise have been found.",
+                )
+                break
+
+
+def strings_to_concatenate(
+    source_path: str,
+) -> Generator[Tuple[str, int, str], None, None]:
+    """
+    This test case is necessary after 'Black' (https://github.com/psf/black),
+    is formating strings over multiple lines.
+
+    For example, when this:
+
+    >>> foo = (
+    ...     "bar "
+    ...     "baz"
+    ... )
+
+    Is becoming this:
+
+    >>> foo = ("bar " "baz")
+
+    'Black' is not considering this as an
+    issue (see https://github.com/psf/black/issues/1051),
+    so we are checking it here instead.
+
+    Parameters
+    ----------
+    source_path : str
+        File path pointing to a single file.
+
+    Yields
+    ------
+    source_path : str
+        Source file path.
+    line_number : int
+        Line number of unconcatenated string.
+    msg : str
+        Explenation of the error.
+
+    Notes
+    -----
+    GH #30454
+    """
+    with open(source_path, "r") as file_name:
+        tokens: List = list(tokenize.generate_tokens(file_name.readline))
+
+    for current_token, next_token in zip(tokens, tokens[1:]):
+        if current_token.type == next_token.type == token.STRING:
+            yield (
+                source_path,
+                current_token.start[0],
+                (
+                    "String unnecessarily split in two by black. "
+                    "Please merge them manually."
+                ),
+            )
+
+
+def strings_with_wrong_placed_whitespace(
+    source_path: str,
+) -> Generator[Tuple[str, int, str], None, None]:
+    """
+    Test case for leading spaces in concated strings.
+
+    For example:
+
+    >>> rule = (
+    ...    "We want the space at the end of the line, "
+    ...    "not at the beginning"
+    ... )
+
+    Instead of:
+
+    >>> rule = (
+    ...    "We want the space at the end of the line,"
+    ...    " not at the beginning"
+    ... )
+
+    Parameters
+    ----------
+    source_path : str
+        File path pointing to a single file.
+
+    Yields
+    ------
+    source_path : str
+        Source file path.
+    line_number : int
+        Line number of unconcatenated string.
+    msg : str
+        Explenation of the error.
+    """
+
+    def get_prefix_len(line: str) -> int:
+        """
+        Get the prefix length from a string literal.
+
+        Parameters
+        ----------
+        line : str
+            Line to check.
+
+        Returns
+        -------
+        int
+            Length of the prefix of the string literal.
+
+        Examples
+        --------
+
+        >>> get_prefix_len('"Hello world"')
+        0
+
+        >>> get_prefix_len('f"Hello world"')
+        1
+
+        >>> get_prefix_len('rf"Hello world"')
+        2
+        """
+        for counter, character in enumerate(line):
```

Comment:
I don't think you need to loop, you can use:
```python
>>> my_string = "rf'foo'"
>>> my_string.index("'")
2

```

Reply:
The two reasons I'm using a loop are:

* I don't know the length of the prefix (If even has a prefix)

* I don't know the correct character that the string is wrapped in (I'll explain):

I can't predict what the character will be, means that I can't know if the string is represented like this:

```python
my_str = "'Hello world'"
```

or like that:

```python
my_str = '"Hello world"'
```

Reply:
I see, but I still think that something like:
```python
min(line.find(quote) for quote in ('"', "'") if line.find(quote) >= 0)
```
should return what you want without having to iterate over the string.

Reply:
__WoW__.
you just blew me away, that's super smart!
I am for real speechless, that's so smart.